src/codegeneration/ArrowFunctionTransformer.js

```diff
@@ -46,6 +59,68 @@ export class ArrowFunctionTransformer extends TempVarTransformer {
    * function body and return statement if needed.
    */
   transformArrowFunctionExpression(tree) {
+    if (this.inDerivedClass_ && this.inConstructor_) {
+      return this.transformUsingCommaExpression_(tree);
+    }
+    return this.transformUsingTempVar_(tree);
+  }
+
+  // This transforms using a comman expression to prevent reading this
+  // before the current location.
+  //
+  //   ($tmp = this, function() {})
+  transformUsingCommaExpression_(tree) {
+    let finder = new FindThisOrArguments();
+    let argumentsTempName, thisTempName;
+    finder.visitAny(tree);
+    if (finder.foundArguments) {
+      argumentsTempName = this.addTempVar();
+      tree = AlphaRenamer.rename(tree, ARGUMENTS, argumentsTempName);
+    }
+    if (finder.foundThis) {
+      thisTempName = this.addTempVar();
+      tree = AlphaRenamer.rename(tree, THIS, thisTempName);
+    }
+
+    let parameterList = this.transformAny(tree.parameterList);
+
+    let body = this.transformAny(tree.body);
+    body = convertConciseBody(body);
+    let functionExpression = new FunctionExpression(tree.location, null,
+        tree.functionKind, parameterList, null, [], body);
+
+    var expressions = [];
+    if (argumentsTempName) {
+      expressions.push(createAssignmentExpression(
+          createIdentifierExpression(argumentsTempName),
+          createIdentifierExpression(ARGUMENTS)));
+    }
+    if (thisTempName) {
+      expressions.push(createAssignmentExpression(
+          createIdentifierExpression(thisTempName),
+          createThisExpression()));
+    }
+
+    if (expressions.length === 0) {
+      return createParenExpression(functionExpression);
+    }
+
+    expressions.push(functionExpression);
+    return createParenExpression(createCommaExpression(expressions));
+  }
+
+  // This transforms the arrow function into:
+  //
+  //   var $tmp = this;
+  //   ...
+  //   (function() {})
+  //
+  // with a temp variable added to the top of the current function.
```

Comment:
Why can't we use the comma expression form? To avoid introducing more temps for 'this'?